turbine/src/quic_endpoint.rs

```diff
@@ -560,6 +596,175 @@ impl<T> From<crossbeam_channel::SendError<T>> for Error {
     }
 }
 
+#[derive(Default)]
+struct TurbineQuicStats {
+    connect_error_invalid_remote_address: AtomicU64,
+    connect_error_other: AtomicU64,
+    connect_error_too_many_connections: AtomicU64,
+    connection_error_application_closed: AtomicU64,
+    connection_error_connection_closed: AtomicU64,
+    connection_error_locally_closed: AtomicU64,
+    connection_error_reset: AtomicU64,
+    connection_error_timed_out: AtomicU64,
+    connection_error_transport_error: AtomicU64,
+    connection_error_version_mismatch: AtomicU64,
+    invalid_identity: AtomicU64,
+    router_try_send_error_full: AtomicU64,
+    send_datagram_error_connection_lost: AtomicU64,
+    send_datagram_error_too_large: AtomicU64,
+    send_datagram_error_unsupported_by_peer: AtomicU64,
+}
+
+async fn report_metrics_task(name: &'static str, stats: Arc<TurbineQuicStats>) {
+    loop {
+        tokio::time::sleep(Duration::from_secs(2)).await;
+        report_metrics(name, &stats);
+    }
+}
+
+fn record_error(err: &Error, stats: &TurbineQuicStats) {
+    match err {
+        Error::CertificateError(_) => (),
```

Comment:
I see we do nothing for several cases; do we just not really expect these to happen at all? Even if we don't break them all out individually, should we make a "other" case as a catch-all so do account for all possible errors?

Reply:
* `CertificateError`, `TlsError` and  `IoError` can only happen during start-up and are passed up the callstack which causes panic: https://github.com/solana-labs/solana/blob/8e0acf481/core/src/validator.rs#L1175-L1189
* `ChannelSendError` only happens during shutdown where downstream processes have winded down and the receiver end of the channel is dropped.
* `SendDatagramError::Disabled` can never happen because that is a local config and we obviously are not disabling datagrams when starting the endpoint.

Reply:
Got it, thanks for the explanation